index.rst

```diff
@@ -1,57 +1,236 @@
+While waiting for those exposures, pipeline-specific calibration data and association data are preloaded into a newly-created handler-local repo.
+As exposures arrive, they are ingested into the local repo, including copying of the image data.
+When the last exposure has arrived, a pipeline selected by one or more fields in the ``next_visit`` event is executed via the Middleware ``pipetask`` command.
+
+The pipeline is responsible for publication of all outputs including Alerts and quality metrics via non-Butler means (although Butler datasets in the handler-local repo can be used as intermediates).
+
+Rationale
+=========
+
+Standard Web API infrastructure provides mechanisms for reliably, elastically handling HTTP verbs.
+Google Cloud Run and AWS Fargate are "serverless" container deployment products that are designed for this, but Kubernetes with ReplicaSet deployments and auto-scaling can be used to do this on-premises.
+Taking advantage of this infrastructure means that all of the resiliency and elasticity requirements, including coordination, distribution, failover, monitoring, etc., can be left to the infrastructure and not custom-built for the application.
+
+Selecting this method of invoking the worker has limitations, however, as the payload to be executed is not a simple stateless one.
+Instead, it needs to receive one or more additional events indicating image arrival.
+This is best sent using a reliable, yet low-latency, mechanism other than HTTP, as listening for HTTP events in an HTTP event handler is tricky.
+AMQP, MQTT, Redis, or possibly Kafka are potential mechanisms here, in addition to Google Pub/Sub or Amazon SQS.
+The trigger for image arrival could be sent by the Camera Control System after it believes it has successfully published the image, or it could be sent by the storage system.
+
+The four LSST Middleware interactions needed by the framework are:
+
+ * On startup, initialize a handler-local Butler repo and copy globally-used calibrations.
+ * On receipt of ``next_visit``, copy calibrations, templates, and potentially APDB data required by the pipeline into the handler-local repo.
+ * On receipt of an image, ingest its metadata.
+ * On receipt of the last image in a visit, trigger execution of the selected pipeline.
+
+Creating a handler-local repo is necessary in order to execute unmodified pipelines, without requiring the writing of special code to work within Prompt Processing.
+This repo (which is expected to be located on a RAM disk or fast local SSD) is used to ensure that all information needed by the pipeline is preloaded to the worker, avoiding bottlenecks during the critical time period after images have been received.
+The code to preload data could be custom or it could be written as PipelineTasks and executed as an "initial step" pipeline, writing outputs to the handler-local repo.
```

Comment:
I don't think a preloading PipelineTask is practical, as a PipelineTask must be able to read input from the repo it runs on and cannot (should not?) do any other I/O. The current `ap_verify` dataset management system performs these tasks using the `butler` command-line utility or the equivalent Python API calls.

Reply:
It's true that PipelineTask is not currently designed to work with multiple Butler repos (at least by default), but I'm not sure it's impossible for it to do so.  The middleware should be able to do things like compare data products between Data Releases, which seems likely to require access to multiple Butler repos.